### PR TITLE
Partial slip boundary conditions

### PR DESCRIPTION
**Pull request by voferreira**
# Description of the problem

- Either slip or noslip boundary conditions return a way thick boundary layer in coarse mesh simulations such as in unresolved CFD-DEM

# Description of the solution

- This new boundary condition allows for an intermediary approach, in which you can set a penalty coefficient `beta` to the normal component and a `boundary layer thickness` to find the tangential component of the velocity at the boundaries.

# How Has This Been Tested?

- Fluid passing a cylinder in 2D.
- Liquid Fluidized Bed in 3d (first 100 timesteps).

# Documentation

- Add the partial slip boundary condition to the parameters guide.
- Add the boundary layer thickness parameter definition, as well.

# Future changes

# Comments

### CODE REVIEW

**Comment by blaisb** [comment]
CI is failing :). Can you also add the boundary condition to the Lethe documentation (modifying the rst files)

File: include/core/parameters.h [hunk 1]
```diff
@@ -954,6 +954,9 @@ namespace Parameters
     // Initial beta
     double beta_0;
 
+    // Initial boundary layer thickness
+    double boundary_layer_thickness_0;
```
Comment (blaisb) [inline include/core/parameters.h:958]: why the _0?

File: include/solvers/navier_stokes_assemblers.h [hunk 2]
```diff
@@ -556,6 +556,52 @@ class WeakDirichletBoundaryCondition : public NavierStokesAssemblerBase<dim>
   const BoundaryConditions::NSBoundaryConditions<dim> &boundary_conditions;
 };
 
+/**
+ * @brief Class that assembles the special case of slip condition using the weak formulation of a Dirichlet boundary condition using the Nitsche method.
+ * This class assembles the weak form of: (u_ib-u)-(u,grad v)
```
Comment (blaisb) [inline include/solvers/navier_stokes_assemblers.h:561]: That's not exactly true right, it assembles partial slip no?

File: source/core/parameters.cc [hunk 3]
```diff
@@ -1996,6 +1996,11 @@ namespace Parameters
                         "0",
                         Patterns::Double(),
                         "Beta coefficient value for the first step time");
+      prm.declare_entry(
+        "initial boundary layer thickness",
+        "0.0000000001",
+        Patterns::Double(),
+        "Bouyndary layer used to calculate the tangent beta coefficient value for the first step time");
```
Comment (blaisb) [inline source/core/parameters.cc:2003]:
```suggestion
        "Bouyndary layer used to calculate the tangential beta coefficient value for the first step time");
```

**Review by blaisb — CHANGES_REQUESTED** [review]
Looks good to me. Just ensure the documentation is done as well.
Could you add a unit test that use this BC? something simple. like a 2D poiseuille flow but with 1 wall with partial slip and the other one no slip. This way it will be easy to see and maintain the BC.

File: source/solvers/navier_stokes_assemblers.cc [hunk 4]
```diff
@@ -1861,6 +1861,230 @@ WeakDirichletBoundaryCondition<dim>::assemble_rhs(
...
+                                {
+                                  for (const unsigned int j :
+                                       scratch_data.fe_face_values
+                                         .dof_indices())
+                                    {
+                                      const auto comp_j =
+                                        fe.system_to_component_index(j).first;
+                                      if (comp_i == comp_j)
+                                        {
+                                          double beta_terms_normal =
```
Comment (Luckabarbeau) [inline source/solvers/navier_stokes_assemblers.cc:1934]: I would also add the normal component of the grad_phi term and the  surface_stress_term
Matrix term
`double grad_phi_terms =
  scratch_data.face_normal[f][q][comp_i]*((viscosity *
    scratch_data
      .face_phi_u[f][q][j]) *
   (scratch_data
      .face_grad_phi_u[f][q][i] *
    scratch_data.face_normal[f][q])) *
  JxW;
double surface_stress_term =
  scratch_data.face_normal[f][q][comp_i]* (viscosity *
   scratch_data
     .face_grad_phi_u[f][q][j] *
   scratch_data.face_normal[f][q]) *
  scratch_data.face_phi_u[f][q][i] *
  JxW;`
RHS
`double grad_phi_terms =
  scratch_data.face_normal[f][q][comp_i]*((viscosity *
    (scratch_data.face_velocity_values[f][q] -
     prescribed_velocity_values[f][q])) *
   (scratch_data.face_grad_phi_u[f][q][i] *
    scratch_data.face_normal[f][q])) *
  JxW;
double surface_stress_term =
  scratch_data.face_normal[f][q][comp_i]*(viscosity *
   scratch_data
     .face_velocity_gradients[f][q] *
   scratch_data.face_normal[f][q] *
   scratch_data.face_phi_u[f][q][i]) *
  JxW;double grad_phi_terms =
  scratch_data.face_normal[f][q][comp_i]*((viscosity *
    (scratch_data.face_velocity_values[f][q] -
     prescribed_velocity_values[f][q])) *
   (scratch_data.face_grad_phi_u[f][q][i] *
    scratch_data.face_normal[f][q])) *
  JxW;
double surface_stress_term =
  scratch_data.face_normal[f][q][comp_i]*(viscosity *
   scratch_data
     .face_velocity_gradients[f][q] *
   scratch_data.face_normal[f][q] *
   scratch_data.face_phi_u[f][q][i]) *
  JxW;`

File: doc/source/parameters/cfd/boundary_conditions_cfd.rst [hunk 5]
```diff
@@ -5,6 +5,7 @@ This subsection defines the boundary conditions associated with fluid dynamics p
 
 * ``noslip`` boundary conditions strongly impose the velocity on a boundary to be :math:`\mathbf{u}=[0,0]^T` and :math:`\mathbf{u}=[0,0,0]^T` in 2D and 3D respectively.
 * ``slip`` boundary conditions impose :math:`\mathbf{u} \cdot \mathbf{n}=0`, with :math:`\mathbf{n}` the normal vector of the boundary. Imposing slip boundary conditions strongly is not trivial in FEM. We refer the reader to the deal.II `documentation <https://www.dealii.org/current/doxygen/deal.II/group__constraints.html>`_ for explanations on how this is achieved.
+* ``partial slip`` boundary condition  introduces the "penalization factor" ``beta`` to the :math:`\mathbf{n}` normal vector of the boundary, and the ``boundary layer thickness`` (m) as a partameter to calculate the shear stress at the boundaries.
```
Comment (Luckabarbeau) [inline doc/source/parameters/cfd/boundary_conditions_cfd.rst:8]:
```suggestion
* ``partial slip`` boundary condition introduces the "penalization factor" ``beta`` to the:math:`\mathbf{n}` normal vector of the boundary, and the ``boundary layer thickness`` (m) as a parameter to calculate the shear stress at the boundaries.
```

**Review by Luckabarbeau — COMMENTED** [review]
Very nice work this will be helpful in future work as well! only a few small comments on the formulation.

**Review by blaisb — APPROVED** [review]
Good to go on my side. Good job, this was not an easy PR to make. Very nice feature.

File: doc/source/parameters/cfd/boundary_conditions_cfd.rst [hunk 6]
```diff
@@ -5,6 +5,7 @@ This subsection defines the boundary conditions associated with fluid dynamics p
 
 * ``noslip`` boundary conditions strongly impose the velocity on a boundary to be :math:`\mathbf{u}=[0,0]^T` and :math:`\mathbf{u}=[0,0,0]^T` in 2D and 3D respectively.
 * ``slip`` boundary conditions impose :math:`\mathbf{u} \cdot \mathbf{n}=0`, with :math:`\mathbf{n}` the normal vector of the boundary. Imposing slip boundary conditions strongly is not trivial in FEM. We refer the reader to the deal.II `documentation <https://www.dealii.org/current/doxygen/deal.II/group__constraints.html>`_ for explanations on how this is achieved.
+* ``partial slip`` boundary condition introduces the "penalization factor" ``beta`` to the:math:`\mathbf{n}` normal vector of the boundary, and the ``boundary layer thickness`` (m) as a parameter to calculate the shear stress at the boundaries.
```
Comment (CaroleAnneDaunais) [inline doc/source/parameters/cfd/boundary_conditions_cfd.rst:8]:
```suggestion
* ``partial slip`` boundary condition introduces the "penalization factor" ``beta`` to the :math:`\mathbf{n}` normal vector of the boundary, and the ``boundary layer thickness`` (m) as a parameter to calculate the shear stress at the boundaries.
```

Comment (CaroleAnneDaunais) [reply doc/source/parameters/cfd/boundary_conditions_cfd.rst:8]: I don't think I would talk about the beta parameter before introducing the equation. I think I would just say that you and manually changing the size of the boundary layer.

File: doc/source/parameters/cfd/boundary_conditions_cfd.rst [hunk 7]
```diff
@@ -75,5 +76,7 @@ where `beta` is a constant and  :math:`(\mathbf{u}\cdot n)_{-}` is :math:`min (0
...
 
-    * ``beta`` is a penalization paramter that is used for both the ``outlet`` and the ``function weak`` boundary conditions. For the outlet boundary conditions, ``beta`` should be close to unity whereas values of ``beta`` of 10 or a 100 can be appropriate for the ``function weak`` boundary condition.
+    * ``beta`` is a penalization parameter used for both the ``outlet``, ``partial slip``, and ``function weak`` boundary conditions. For the outlet boundary conditions ``beta`` should be close to unity, whereas ``beta`` of 10 or a 100 can be appropriate for the ``function weak`` boundary condition. For the ``partial slip`` condition, use high values of ``beta`` (i.e., > 50).
+
+    * ``boundary layer thickness`` (:math:`d_w`) is the parameter applied to the ``partial slip`` boundary condition. It is used to estimate the tangential shear stress :math:`\tau_t = -\mu \frac{u}{d_w}`. For very high ``boundary layer thicknes``, the boundary layer should behave exactly like the ``slip`` condition.
```
Comment (CaroleAnneDaunais) [inline doc/source/parameters/cfd/boundary_conditions_cfd.rst:81]: You should add this parameter to the boundary conditions subsection, with it's default value :)

File: source/solvers/navier_stokes_assemblers.cc [hunk 8]
```diff
@@ -1861,6 +1861,276 @@ WeakDirichletBoundaryCondition<dim>::assemble_rhs(
...
+                                                                                        scratch_data.face_normal[f][q])) *
+                                                                                      JxW;
+                                                                                    double surface_stress_term =
+                                                                                      scratch_data.face_normal[f][q][comp_i]* (viscosity *
+                                                                                       scratch_data
+                                                                                         .face_grad_phi_u[f][q][j] *
+                                                                                       scratch_data.face_normal[f][q]) *
+                                                                                      scratch_data.face_phi_u[f][q][i] *
+                                                                                      JxW;
+                                          */
```
Comment (CaroleAnneDaunais) [inline source/solvers/navier_stokes_assemblers.cc:1980]: should this part be erased? Or do you want to keep it for later?

**Review by CaroleAnneDaunais — COMMENTED** [review]
Very nice Victor! Apart from my minor comments, this is good to go for me.

File: source/solvers/navier_stokes_assemblers.cc (hunk 8 above)
Comment (voferreira) [reply source/solvers/navier_stokes_assemblers.cc:1980]: It should be kept there for now.

**Review by CaroleAnneDaunais — APPROVED** [review]

**Pull request merged**